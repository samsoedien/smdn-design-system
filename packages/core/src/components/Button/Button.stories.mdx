import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks'
import Button from './index'

<Meta title="Misc" />

Here's some _markdown_!

# Preview

<Preview>
  <Story name="MDX Button">
    <Button>Hello World</Button>
  </Story>
</Preview>

# Props

<Props of={Button} />
