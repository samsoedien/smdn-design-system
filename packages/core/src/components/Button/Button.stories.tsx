import React from 'react'
import { action } from '@storybook/addon-actions' // eslint-disable-line

import Button from './index'

export default {
  title: 'Core/Button',
  component: Button,
  componentTitle: 'Basic buttons to provide the user with an action when user interaction is required.',
  decorators: [],
}

export const ButtonDefault = () => <Button onClick={action('button action click')}>Button</Button>

ButtonDefault.story = {
  name: 'Button',
  parameters: {
    jest: ['Button.test.tsx', 'Button.snapshot.test.tsx'],
    // jest: { disable: true },
  },
}

export const ButtonSizes = () => (
  <>
    <Button size="small" onClick={action('button action click')}>
      Small Button
    </Button>
    <Button size="regular" onClick={action('button action click')}>
      Regular Button
    </Button>
    <Button size="large" onClick={action('button action click')}>
      Large Button
    </Button>
  </>
)

ButtonSizes.story = {
  name: 'Button - Sizes',
  parameters: {
    jest: ['Button.test.tsx', 'Button.snapshot.test.tsx'],
    // jest: { disable: true },
  },
}

export const ButtonAppearances = () => (
  <>
    <Button appearance="primary" onClick={action('button action click')}>
      Primary Button
    </Button>
    <Button appearance="secondary" onClick={action('button action click')}>
      Secondary Button
    </Button>
    <Button appearance="tertiary" onClick={action('button action click')}>
      Tertiary Button
    </Button>
  </>
)

ButtonAppearances.story = {
  name: 'Button - Appearances',
  parameters: {
    jest: ['Button.test.tsx', 'Button.snapshot.test.tsx'],
    // jest: { disable: true },
  },
}

export const ButtonLoading = () => <Button isLoading>Loading Button...</Button>

ButtonLoading.story = {
  name: 'Button - Loading',
  parameters: {
    jest: ['Button.test.tsx', 'Button.snapshot.test.tsx'],
    // jest: { disable: true },
  },
}
