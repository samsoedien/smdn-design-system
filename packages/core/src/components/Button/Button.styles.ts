import styled, { css } from 'styled-components'

import { IButtonProps } from './Button'

export const StyledButton = styled.button<IButtonProps>`
  border: none;
  border-radius: 3px;
  padding: 0 12px;
  background-color: lightgrey;
  transition: background-color 0.2s;

  &:hover {
    background-color: darkgrey;
    cursor: pointer;
  }

  ${(props) =>
    props.size === 'regular' &&
    css`
      height: 36px;
    `};

  ${(props) =>
    props.size === 'small' &&
    css`
      height: 30px;
    `};

  ${(props) =>
    props.size === 'large' &&
    css`
      height: 48px;
    `};

  ${(props) =>
    props.appearance === 'primary' &&
    css`
      background-color: teal;
      color: white;
    `};

  ${(props) =>
    props.appearance === 'secondary' &&
    css`
      color: #121212;
    `};

  ${(props) =>
    props.appearance === 'tertiary' &&
    css`
      background-color: transparent;
      color: #121212;
    `};

  ${(props) =>
    props.isLoading &&
    css`
      &:hover {
        cursor: progress;
      }
    `}
`
