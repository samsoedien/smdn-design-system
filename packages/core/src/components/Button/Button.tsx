import React from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'

import { Appearance } from '@smdn-design-system/shared'

const StyledButton = styled.button<IButtonProps>`
  ${(props) =>
    props.appearance === 'primary' &&
    css`
      background-color: blue;
    `};
  ${(props) =>
    props.appearance === 'secondary' &&
    css`
      background-color: grey;
    `};
`

export interface IButtonProps {
  appearance?: Appearance
  onClick?: () => void
  children: React.ReactNode
}

const Button: React.FC<IButtonProps> = ({ appearance, children }) => (
  <StyledButton appearance={appearance} className="smdn-button" data-test="button-component">
    {children}
  </StyledButton>
)

Button.defaultProps = {
  appearance: 'secondary',
}

Button.propTypes = {
  appearance: PropTypes.oneOf<Appearance>(['primary', 'secondary', 'tertiary']),
  children: PropTypes.node.isRequired,
}

export default Button
