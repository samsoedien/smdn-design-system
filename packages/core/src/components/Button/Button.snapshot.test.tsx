import React from 'react'
import styled from 'styled-components'
import renderer from 'react-test-renderer' // eslint-disable-line
import 'jest-styled-components'

import Button from './Button'

test('Button snapshot works', () => {
  const tree = renderer.create(<Button>Button</Button>).toJSON()
  expect(tree).toMatchSnapshot()
})
