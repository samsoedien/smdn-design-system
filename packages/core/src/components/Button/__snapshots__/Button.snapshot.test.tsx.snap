// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshot works 1`] = `
<button
  className="smdn-b-button"
  data-test="button-component"
  disabled={false}
  onClick={[Function]}
  size="regular"
  type="submit"
>
  Button
</button>
`;
